pkg/function/restore_data_using_kopia_server.go

```diff
@@ -0,0 +1,261 @@
+package function
+
+import (
+	"context"
+	"fmt"
+
+	crv1alpha1 "github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	kanister "github.com/kanisterio/kanister/pkg"
+	"github.com/kanisterio/kanister/pkg/format"
+	kankopia "github.com/kanisterio/kanister/pkg/kopia"
+	kopiacmd "github.com/kanisterio/kanister/pkg/kopia/command"
+	"github.com/kanisterio/kanister/pkg/kube"
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+const (
+	RestoreDataUsingKopiaServerFuncName = "RestoreDataUsingKopiaServer"
+	// SparseRestoreOption is the key for specifiying whether to do a sparse restore
+	SparseRestoreOption = "sparseRestore"
+)
+
+type restoreDataUsingKopiaServerFunc struct{}
+
+func init() {
+	err := kanister.Register(&restoreDataUsingKopiaServerFunc{})
+	if err != nil {
+		return
+	}
+}
+
+var _ kanister.Func = (*restoreDataUsingKopiaServerFunc)(nil)
+
+func (*restoreDataUsingKopiaServerFunc) Name() string {
+	return RestoreDataUsingKopiaServerFuncName
+}
+
+func (*restoreDataUsingKopiaServerFunc) RequiredArgs() []string {
+	return []string{
+		RestoreDataBackupIdentifierArg,
+		RestoreDataNamespaceArg,
+		RestoreDataRestorePathArg,
+		RestoreDataImageArg,
+	}
+}
+
+func (*restoreDataUsingKopiaServerFunc) Arguments() []string {
+	return []string{
+		RestoreDataBackupIdentifierArg,
+		RestoreDataNamespaceArg,
+		RestoreDataRestorePathArg,
+		RestoreDataPodArg,
+		RestoreDataVolsArg,
+		RestoreDataPodOverrideArg,
+		RestoreDataImageArg,
+	}
+}
+
+func (*restoreDataUsingKopiaServerFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]any) (map[string]any, error) {
+	var (
+		err         error
+		image       string
+		namespace   string
+		restorePath string
+		snapID      string
+	)
+	if err = Arg(args, RestoreDataBackupIdentifierArg, &snapID); err != nil {
+		return nil, err
+	}
+	if err = Arg(args, RestoreDataNamespaceArg, &namespace); err != nil {
+		return nil, err
+	}
+	if err = Arg(args, RestoreDataRestorePathArg, &restorePath); err != nil {
+		return nil, err
+	}
+	if err = Arg(args, RestoreDataImageArg, &image); err != nil {
+		return nil, err
+	}
+
+	userPassphrase, cert, err := userCredentialsAndServerTLS(&tp)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch User Credentials/Certificate Data from Template Params")
+	}
+
+	fingerprint, err := kankopia.ExtractFingerprintFromCertificateJSON(cert)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch Kopia API Server Certificate Secret Data from Certificate")
+	}
+
+	// Validate and get optional arguments
+	pod, vols, podOverride, err := validateAndGetOptArgsForRestore(tp, args)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vols) == 0 {
+		vols, err = FetchPodVolumes(pod, tp)
```

Comment:
Since Pod name is optional, what happens if its empty?

Reply:
So we use `Pod` option if we want to Restore Data backupped by our Backupdatausingkopia function and we use `Volumes` if we want to restore data backupped by Copyvolumedatausingkopiaserver function

Reply:
Either one of them should be present if we want to use this function